Add TablePathResolver classes

AnanaMJ opened the pull request:
This was suggested to be done after these classes from [housekeeping](https://github.com/HotelsDotCom/housekeeping) were needed in another project.

AnanaMJ committed:
Add TablePathResolver classes

massdosage commented on src/main/java/com/hotels/hcommon/hive/metastore/paths/PartitionedTablePathResolver.java:
```diff
@@ -0,0 +1,98 @@
+    tableBaseLocation = locationAsPath(table);
+    List<Partition> onePartition = metaStore.listPartitions(table.getDbName(), table.getTableName(), (short) 1);
+    if (onePartition.isEmpty()) {
+      String errorMessage = String.format("Table '%s' has no partitions, perhaps you can simply delete: %s.",
```
Since this is now being used outside of housekeeping it probably doesn't make sense to talk about deletion. The error message should just indicate that it has no partitions (i.e. WTF are you doing trying to use this partitioned resolver on a table with no partitions).

AnanaMJ replied:
Agree, although it would be the factory deciding what class to use, so if we get to that error message, the table has partition keys, but they're empty. Can that happen?

massdosage replied:
I guess the `getPartitionKeys()` (i.e. "yes this table is partitioned") could return something but `listPartitions` could return empty (i.e. "but none have been added"). Edge case, not sure how useful it is but let's just keep this for now as its backwards compatible and we change it in future if we decide we don't need this behaviour here (in which case we'd probably need to move the check into the vacuum tool if it was the only place that actually needed it).

massdosage commented on src/main/java/com/hotels/hcommon/hive/metastore/paths/PartitionedTablePathResolver.java:
```diff
@@ -0,0 +1,98 @@
+      throws NoSuchObjectException, MetaException, TException {
+    this.metaStore = metaStore;
+    this.table = table;
+    log.debug("Table '{}' is partitioned", Warehouse.getQualifiedName(table));
```
Not sure this adds much value, remove?

massdosage commented on src/main/java/com/hotels/hcommon/hive/metastore/paths/PartitionedTablePathResolver.java:
```diff
@@ -0,0 +1,98 @@
+    if (onePartition.isEmpty()) {
+      String errorMessage = String.format("Table '%s' has no partitions, perhaps you can simply delete: %s.",
+          Warehouse.getQualifiedName(table), tableBaseLocation);
+      log.warn(errorMessage);
```
Usually if we throw the exception we don't log it since we don't need the message showing up twice so I think we should remove the warn

massdosage commented on src/main/java/com/hotels/hcommon/hive/metastore/paths/PartitionedTablePathResolver.java:
```diff
@@ -0,0 +1,98 @@
+          .startsWith(tableBaseLocation.toString().toLowerCase(Locale.ROOT))) {
+        String errorMessage = String.format("Check your configuration: '%s' does not appear to be part of '%s'.",
+            location, tableBaseLocation);
+        log.error(errorMessage);
```
Similar here, we shouldn't log if we're going to throw since the exception should end up in the logs anyway.

massdosage commented on src/main/java/com/hotels/hcommon/hive/metastore/paths/TablePathResolver.java:
```diff
@@ -0,0 +1,49 @@
+        IMetaStoreClient metastore, Table table)
+        throws URISyntaxException, NoSuchObjectException, MetaException, TException {
+      if (!table.getPartitionKeys().isEmpty()) {
+        return new PartitionedTablePathResolver(metastore, table);
```
If we did want to log whether the table was partitioned or not I'd say this if statement would be the place to do it.

massdosage commented on src/main/java/com/hotels/hcommon/hive/metastore/paths/UnpartitionedTablePathResolver.java:
```diff
@@ -0,0 +1,62 @@
+  UnpartitionedTablePathResolver(Table table) {
+    this.table = table;
+    Path localTableBaseLocation = locationAsPath(table);
+    log.debug("Table '{}' is unpartitioned", Warehouse.getQualifiedName(table));
```
I'd suggest removing this entirely or moving up into the `Factory` above (see previous comment).

massdosage reviewed (changes requested):
Looks most fine, just a few comments from me about what we log and where now that's in a more generic location.

AnanaMJ committed:
Remove some logs and update minor version

AnanaMJ commented:
@massdosage I've updated the PR with your suggested changes (moved the log to the factory because sometimes it's useful seeing that the table is picked up correctly) and I updated the minor version since it's something added, but in a backwards compatible way :)
